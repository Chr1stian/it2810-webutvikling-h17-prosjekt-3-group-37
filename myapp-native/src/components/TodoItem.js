import React from 'react';
import { Button, StyleSheet, Text, View } from 'react-native';

export default class TodoItem extends React.Component {

  deleteTodoItem = () => {
    this.props.deleteTodoItem(this.props.todoItem);
  }

  setStatusTodoItem = () => {
    this.props.setStatusTodoItem(this.props.todoItem);
  }


  render() {

    let { todoItem } = this.props;

    let statusText = todoItem.finished ? "Undo" : "Done";
    let backgroundColor = todoItem.finished ? "#007a87" : "#555555";
    let buttonColor = todoItem.finished ? "#0097a7" : "#303030";

    const styles = StyleSheet.create({
      container: {
        flexDirection: 'row',
        justifyContent: 'space-between',
        margin: 5,
        marginTop: 10,
        backgroundColor: backgroundColor,
        borderRadius: 1,
        paddingTop: 5,
        paddingBottom: 5,
        paddingLeft: 5,
        paddingRight: 5,
      },

      buttonContainer: {
        marginLeft: 5,
      },

      buttons: {
        flexDirection: 'row',
        margin: 5,
      },

      textContainer: {
        flexDirection: 'column',
        justifyContent: 'space-around',
        margin: 5,
        flex: 1

      },

      title: {
        color: 'white',
        fontWeight: 'bold'

      },

      text: {
        color: 'white',

      }


    });

    return (
      <View style={styles.container}>
        <View style={styles.textContainer}>
          <Text style={styles.title}>{todoItem.title}</Text>
          <Text style={styles.text}>{todoItem.text}</Text>
        </View>
        <View style={styles.buttons}>
          <View style={styles.buttonContainer}>
            <Button
              title={statusText}
              onPress={this.setStatusTodoItem}
              color={buttonColor}

            />
          </View>
          <View style={styles.buttonContainer}>
            <Button
              title="Delete"
              onPress={this.deleteTodoItem}
              color={buttonColor}

              

            />
          </View>
        </View>


      </View>
    );
  }
}
