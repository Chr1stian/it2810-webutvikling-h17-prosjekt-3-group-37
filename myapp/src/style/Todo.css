.inputs {
	display: flex;
	flex-direction: row;
}

.todoItemText {
	display: flex;
	justify-content: space-between;
	flex-grow: 100;

}

.cardText {
	flex-grow: 100;
	text-align: right;
	font-style: italic;
}

.cardHeader {
	flex-grow: 100;
}

.todoItemButtons {
	display: flex;
	flex-direction: row;
	flex-grow: 0;

}

.todoItemCreator {
	margin-top: 20px;
	margin-left: auto;
	margin-right: auto;

}

#toggleDone div {
	margin-left : 20px;
	margin-right: 20px;
	margin-top: 5px;
	margin-bottom: 5px;
}

.todoItemContainer {
	width: 40%;
	margin-right: auto;
	margin-left: auto;
    display: flex;
    flex-direction: column;
    margin-top: 30px;
}

.creatorContainer {
	margin-left: auto;
	margin-right: auto;
	display: flex;
	flex-flow: row wrap;

}

@media screen and (max-width: 1200px){

	.todoItemContainer {
		width: 60%;


	}
}

@media screen and (max-width: 900px){

	.todoItemContainer {
		width: 75%;


	}
}

@media screen and (max-width: 800px){
	.todoItemContainer {
		width: 85%;
	}
}

@media screen and (max-width: 700px){
	.todoItemContainer {
		width: 95%;
	}

	.todoItemCreator {
	}


}

@media screen and (max-width: 630px){
	.todoItemContainer {
		width: 100%;

	}	.todoItemCreator {
		margin-left: 10px;
		margin-right: 10px;
	}



	#toggleDone {
		margin-left: 10px !important;
	}


}