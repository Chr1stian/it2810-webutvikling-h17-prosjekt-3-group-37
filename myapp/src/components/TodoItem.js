//Import React, ReactComponents and style
import React, { Component } from 'react';
import './../style/TodoItem.css';

//Import Material UI
import {Card, CardActions, CardText} from 'material-ui/Card';
import FlatButton from 'material-ui/FlatButton';


class TodoItem extends Component {
	//Passes the item that should be deleted as prop to the Todo.js components deleteTodoItem function
	deleteTodoItem = () => {
		this.props.deleteTodoItem(this.props.todoItem);
	}
	//Passes the item that should have its status changed as prop to the Todo.js components setStatusTodoItem function
	setStatusTodoItem = () => {
		this.props.setStatusTodoItem(this.props.todoItem);
	}

	render() {
		let {todoItem} = this.props;
		//Styles for the todoItems text
		const textStyle = { width: '150px', whiteSpace: 'normal', wordWrap: 'break-word', textWrap: 'wrap'};
		const headerStyle = { width: '75px', whiteSpace: 'normal', wordWrap: 'break-word', textWrap: 'wrap'};


		//Sets status text and colors for each card
		let status = todoItem.finished ? "Undone" : "Done";
		let buttonColor = todoItem.finished ? "#007a87" : "#555555";
		let cardColor = todoItem.finished ? "#0097a7" : "#303030";

		return (
      	//Creates and returns the todoItem card to the Todo.js Component
			<div>
				<Card style={{backgroundColor: cardColor}} className="TodoItem">
					<div>
						<div className="TodoItemText">
							<CardText className="CardHeader" style={headerStyle}>{todoItem.title}</CardText>
							<CardText className="CardText" style={textStyle} >{ todoItem.text }</CardText>
						</div>
						<CardActions className="TodoItemButtons">
							<FlatButton style={{backgroundColor: buttonColor}} id="DoneButton" onClick = {this.setStatusTodoItem}>{status}</FlatButton>
							<FlatButton style={{backgroundColor: buttonColor}} onClick = {this.deleteTodoItem}>Delete</FlatButton>
						</CardActions>
					</div>
				</Card>
			</div>
		)
	}
}
export default TodoItem;
