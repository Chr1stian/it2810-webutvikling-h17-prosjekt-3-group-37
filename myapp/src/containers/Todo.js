//Import React and ReactComponents, style and components used in this component
import React, { Component } from 'react';
import TodoItem from './../components/TodoItem';
import './../style/Todo.css';

//Import Material UI
import {Card, CardActions, CardText} from 'material-ui/Card';
import FlatButton from 'material-ui/FlatButton';
import TextField from 'material-ui/TextField';

//Import RandomID-generator
import uuid from 'uuid';

class Todo extends Component {
	constructor(props){
		super(props);
		this.state = {
			todolist: []
		}
	}
	//Sets the states todolist from the users localStorage
	componentWillMount = () => {
	    let todolist = JSON.parse(localStorage.getItem('todoitems'));
	    this.setState({
	        todolist: todolist || []
	    })
	}
	//Creates a todo item and adds it to the list
	addTodoItem = () => {
		let title = document.getElementById("TitleField").value;
		let text = document.getElementById("CommentField").value;
		if (title !== "") {
			let {todolist} = this.state;
			//Adds the new todo item to the list
			todolist.push({ID: uuid.v4(), title: title, text: text, finished: false});
			this.setState({todolist: todolist});
			//Saves the list to localStorage
			localStorage.setItem('todoitems', JSON.stringify(todolist));
			//Resets the new Todo item fields and hint text
			document.getElementById("TitleField").value = "";
			document.getElementById("CommentField").value = "";
			document.getElementById("TitleField").placeholder = "What task?";
			document.getElementById("CommentField").placeholder = "Optional comment";
		}
	}
	//Deletes the selected todo item
	deleteTodoItem = (todoItem) => {
		let {todolist} = this.state;
		let i = todolist.indexOf(todoItem);
		todolist.splice(i, 1);
		this.setState({todolist: todolist});
		localStorage.setItem('todoitems', JSON.stringify(todolist));
	}
	//Sets the todo item to either done or undone
	setStatusTodoItem = (todoItem) => {
		let {todolist} = this.state;
		let i = todolist.indexOf(todoItem);
		todolist[i].finished = !todolist[i].finished;
		this.setState({todolist: todolist});
		localStorage.setItem('todoitems', JSON.stringify(todolist));
	}
	//Shows or hides finnished tasks
	toggleDone = () => {
		let setvalue = "None";
		let setText = "Show finished tasks";
		if (document.getElementById("Done").style.display === "none"){
			setvalue = "Flex";
			setText = "Hide finished tasks";
		}
		document.getElementById("Done").style.display = setvalue;
		document.getElementById("ToggleDone").getElementsByTagName("div")[0].innerHTML = setText;
	}
	render() {
		let { todolist } = this.state;
		//Gets a list of items marked as done with the .finished boolean
		let doneTodoList = todolist
			.filter(function (todoItem) {return todoItem.finished})
			.map((todoItem) =>
				<TodoItem
					key={todoItem.ID}
					todoItem={todoItem}
					setStatusTodoItem={this.setStatusTodoItem}
					deleteTodoItem={this.deleteTodoItem}
				/>
			);

		//Displays a button to hide finnished tasks if you have any
		let showTasks;
		if (todolist.length === 0) {
			showTasks = <Card><CardText>You have no items todo..</CardText></Card>;
		}else if (todolist.length === doneTodoList.length) {
			let buttonStyle = {backgroundColor: "#555555", height: "auto", marginTop: "20px", marginBottom: "-20px"}
			showTasks = <div><Card><CardText>You have no items todo..</CardText></Card>
			<FlatButton style={buttonStyle} id="ToggleDone" onClick = {this.toggleDone}>Hide finished tasks</FlatButton></div>;
		}
		 else {
			if (doneTodoList.length !== 0) {
				let buttonStyle = {backgroundColor: "#555555", height: "auto", marginTop: "20px", marginBottom: "-20px"}
				showTasks = <FlatButton style={buttonStyle} id="ToggleDone" onClick = {this.toggleDone}>Hide finished tasks</FlatButton>;
			}
		}

		/*
	      Creates a Material UI card with fields and button to create a new todo item
	      Iterates trough the todolist with all todo items and fills the todoItemCointainer with a card for each
				If there are todo items marked as done, the button "Hide finished tasks" will appear
				The todoItemCointainer with id "done" will be be filled with a card for each
	    */
		return (

     		<div>
	      		<div className="CreatorContainer">
					<Card className="TodoItemCreator">
						<CardActions className="TodoItemCreatorContent">
							<TextField id="TitleField" hintText="What task?"/>
							<TextField id="CommentField" hintText="Optional comment" />
							<FlatButton style={{backgroundColor: "#555555"}} onClick = {this.addTodoItem}>Add</FlatButton>
						</CardActions>
					</Card>
				</div>
				<div className="TodoItemContainer" id="NotDone">
					{ todolist
						.filter(function (todoItem) {return !todoItem.finished})
						.map((todoItem) =>
							<TodoItem
								key={todoItem.ID}
								todoItem={todoItem}
								setStatusTodoItem={this.setStatusTodoItem}
								deleteTodoItem={this.deleteTodoItem}
							/>
						)
					}
					<div>
						{showTasks}
					</div>
				</div>
				<div className="TodoItemContainer" id="Done">
					{doneTodoList}
				</div>
			</div>
		)
	}
}
export default Todo;
