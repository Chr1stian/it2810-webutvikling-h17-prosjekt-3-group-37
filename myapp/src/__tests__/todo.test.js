import React from "react";
//Import Enzyme with necessary functions to test component as a unit
import Adapter from "enzyme-adapter-react-16";
import Enzyme from "enzyme";
Enzyme.configure({ adapter: new Adapter() });
import { shallow, mount, render } from "enzyme";
import toJson from "enzyme-to-json";
//Importing component that will be tested
import Todo from "./../containers/Todo.js";
import TodoItem from "./../components/TodoItem.js";
import "./../test/mock-localstorage.js";

//takes snapshot of component
test("Component renders without crashing", () => {
  const component = shallow(<Todo />);
  //console.log(component);
  let tree = toJson(component);
  expect(tree).toMatchSnapshot();
});

//check what html are being rendered, not really useful but was first test written
test("Checking if a div is being rendered", () => {
  const wrapper = shallow(<Todo />);
  expect(wrapper.type()).toEqual("div");
});

test("Checking if new item in todolist is added to list", () => {
  const wrapper = shallow(<Todo />);
  try {
    const todoitem = shallow(<TodoItem />);
  } catch (e) {
    return undefined;
  } finally {
    const todo = {
      ID: "8d4c43b0-eac3-4af3-bea0-1c2ebd4165cb",
      title: "Dette er en tittel",
      text: "Dette er teksten",
      finished: false
    };
    wrapper.instance().setState(todo);
    expect(wrapper.instance().state).toMatchObject(todo);
  }
});
