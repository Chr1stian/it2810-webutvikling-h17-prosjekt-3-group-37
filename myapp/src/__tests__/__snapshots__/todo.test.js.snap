// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component renders without crashing 1`] = `
<div>
  <div
    className="CreatorContainer"
  >
    <Card
      className="TodoItemCreator"
      expandable={false}
      expanded={null}
      initiallyExpanded={false}
    >
      <CardActions
        className="TodoItemCreatorContent"
      >
        <TextField
          disabled={false}
          floatingLabelFixed={false}
          fullWidth={false}
          hintText="What task?"
          id="TitleField"
          multiLine={false}
          rows={1}
          type="text"
          underlineShow={true}
        />
        <TextField
          disabled={false}
          floatingLabelFixed={false}
          fullWidth={false}
          hintText="Optional comment"
          id="CommentField"
          multiLine={false}
          rows={1}
          type="text"
          underlineShow={true}
        />
        <FlatButton
          disabled={false}
          fullWidth={false}
          labelPosition="after"
          labelStyle={Object {}}
          onClick={[Function]}
          onKeyboardFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onTouchStart={[Function]}
          primary={false}
          secondary={false}
          style={
            Object {
              "backgroundColor": "#555555",
            }
          }
        >
          Add
        </FlatButton>
      </CardActions>
    </Card>
  </div>
  <div
    className="TodoItemContainer"
    id="NotDone"
  >
    <div>
      You have no items todo
    </div>
  </div>
  <div
    className="TodoItemContainer"
    id="Done"
  />
</div>
`;
